dbms/programs/odbc-bridge/Handlers.cpp

```diff
@@ -0,0 +1,153 @@
+#include <common/logger_useful.h>
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int BAD_REQUEST_PARAMETER;
+}
+
+namespace
+{
+    std::unique_ptr<Block> parseColumns(std::string && column_string)
+    {
+        std::unique_ptr<Block> sample_block = std::make_unique<Block>();
+        auto names_and_types = NamesAndTypesList::parse(column_string);
+        for (const NameAndTypePair & column_data : names_and_types)
+            sample_block->insert({column_data.type, column_data.name});
+        return sample_block;
+    }
+
+    size_t parseMaxBlockSize(const std::string & max_block_size_str, Poco::Logger * log)
+    {
+        size_t max_block_size = DEFAULT_BLOCK_SIZE;
+        if (!max_block_size_str.empty())
+        {
+            try
+            {
+                max_block_size = std::stoul(max_block_size_str);
+            }
+            catch (...)
+            {
+                tryLogCurrentException(log);
+            }
+        }
+        return max_block_size;
+    }
+}
+
+
+ODBCHandler::PoolPtr ODBCHandler::getPool(const std::string & connection_str)
+{
+    if (!pool_map->count(connection_str))
+    {
+        std::lock_guard lock(mutex);
+        pool_map->emplace(connection_str, createAndCheckResizePocoSessionPool([connection_str] {
+            return std::make_shared<Poco::Data::SessionPool>("ODBC", connection_str);
+        }));
+    }
+    return pool_map->at(connection_str);
+}
+
+void ODBCHandler::handleRequest(Poco::Net::HTTPServerRequest & request, Poco::Net::HTTPServerResponse & response)
+{
+    Poco::Net::HTMLForm params(request, request.stream());
+    LOG_TRACE(log, "Request URI: " + request.getURI());
+
+    auto process_error = [&response, this](const std::string & message) {
+        response.setStatusAndReason(Poco::Net::HTTPResponse::HTTP_INTERNAL_SERVER_ERROR);
+        if (!response.sent())
+            response.send() << message << std::endl;
+        LOG_WARNING(log, message);
+    };
+
+    if (!params.has("query"))
+    {
+        process_error("ODBCBridge: No 'query' in request body");
+        return;
+    }
+
+    if (!params.has("columns"))
+    {
+        process_error("ODBCBridge: No 'columns' in request URL");
+        return;
+    }
+
+    if (!params.has("connection_string"))
+    {
+        process_error("ODBCBridge: No 'connection_string' in request URL");
+        return;
+    }
+
+    size_t max_block_size = parseMaxBlockSize(params.get("max_block_size", ""), log);
+
+    std::string columns = params.get("columns");
+    std::unique_ptr<Block> sample_block;
+    try
+    {
+        sample_block = parseColumns(std::move(columns));
+    }
+    catch (const Exception & ex)
+    {
+        process_error("ODBCBridge: Invalid 'columns' parameter in request body '" + ex.message() + "'");
+        return;
+    }
+
+    std::string format = params.get("format", "RowBinary");
+    std::string query = params.get("query");
+    LOG_TRACE(log, "ODBCBridge: Query '" << query << "'");
```

Comment:
You put query in single quotes, but query is unescaped - that may lead to confusion, because query often contains single quotes inside.
Better to print query after colon.